### DISCUSSION TITLE
Real LED Resonance Fitting

### DISCUSSION THREAD

**Discussion by TiptoC01** [MilesCranmer/PySR#1115] [Q&A]
Dear All 

For a wider project I am using LED light sources from thor labs 

eg. https://www.thorlabs.com/fiber-coupled-leds?tabName=Overview

I am hoping to do some Bayesian infrerencing with the distributions which are kindle providing as raw data by Thor labs.   Normally I would simply put a smoothing filter through one of the spectra and be on my way.  However, I thought why not try and get an actual mathematical expression from the intensity distributions from PySR to do this.    

It would appear trying to fit any of these uni-modal non guassian curves is very challenging for PySR and I don't really understand why.

By passing removing  very small intensity   readings (i.e a noise floor) from the data,  passing  an empirical mean wavelegth and a emprical standard deviation in addtion to the wavelength and the instensity.   I do finally get a curve that goes though the peak.  but does  terrible job on the noise floor.   If I leave the nose floor in it seems to over fit the noise floor but ignore the peak. 

this is for the LED product number  M530F3  

<img width="839" height="740" alt="image" src="https://github.com/user-attachments/assets/8d149c1a-cdc5-4da4-90ac-cc0349a7a6f8" />

Am I doing this in a wrong headed way?   This is a code snippet

<pre>
testx = data_normalized_sub[["wavelength","lambda_bar","sigma"]][:]
testy = data_normalized_sub["intensity"][:].to_numpy()

model = PySRRegressor(
    maxsize=40,
    niterations=1000,  
    populations=3*8,
    parsimony=1e-6,
    binary_operators=["+","*","^"],
    unary_operators=["exp"], 
    elementwise_loss= "LPDistLoss{2}()",
    constraints={"^":(-1,9)}, # only a constant or a linear term in exponent 
    nested_constraints={
        "^": { "^": 1  },
             "exp": { "exp":0  , "^": 1} # exponention can be 1 deep (e^x^2 is allowed, but not e^e^x)
    },
    model_selection="best"
)

model.fit(testx, testy)
</pre>

If I put the noise floor back in I get this 

<img width="839" height="740" alt="image" src="https://github.com/user-attachments/assets/2a905492-f93b-4b63-96d2-1e2d177ac17c" />

**Comment by MilesCranmer** [comment]
You could try passing a Gaussian operator, like this:
```
"gauss(x) = exp(-x^2)"
```
That might help it get the peak shape.

Perhaps worth trying a skewed Gauss operator as well.

Also it might be worth giving it fake data to nail down the asymptotic behavior. For example you could add a data point (x=1e9, y=0) and weight it very strongly so it definitely passes through that point.

**Comment by MilesCranmer** [comment]
I might also try L1 loss to see?

**Comment by TiptoC01** [comment]
Thank you for the suggestions, Miles.

For completeness these were my conclusions 

After a lot playing around with this I found the best method appeared to be to try and fit the log of the data, only using the data near the peak.    i.e remove the noise floor.   The filtered data on a log scale looks like it might be hyperbola
 
<img width="839" height="740" alt="image" src="https://github.com/user-attachments/assets/ef1fad2d-a8fc-4fdb-9993-0e8cf6c24687" />

So I tried this 

<pre>model = PySRRegressor(
maxsize=25,
    niterations=1000,  
    binary_operators=["+","*"], 
    unary_operators=["sqrt","square"],
    complexity_of_constants=1,
    elementwise_loss="LPDistLoss(2)",
    batching=True,  
    batch_size=128,
    parsimony=1e-10,
    model_selection="score",
)

model.fit(filters[["Wavelength"]], np.log(filters['GreenIntensity'].to_numpy()))
</pre>

Which then does a pretty good job of finding curves that actually go through the data.    

<img width="839" height="740" alt="image" src="https://github.com/user-attachments/assets/364bcaec-feb9-4b6e-b0cd-f19cfc188b46" />

I think the take aways are don't fit noise?  And be careful with exponentials and powers?

**Reply by MilesCranmer** [reply]
Nice!! Great that you got it working
